File: fixers/uses-debhelper-compat-file.py
#!/usr/bin/python3

import os
import sys

from typing import List

from debmutate.control import (
    ensure_exact_version,
    iter_relations,
    is_relation_implied,
    parse_relations,
    format_relations,
    )
from debmutate.debhelper import (
    read_debhelper_compat_file,
    )
from lintian_brush.fixer import control, report_result, fixed_lintian_tag
from lintian_brush.debhelper import highest_stable_compat_level
from debmutate._rules import (
    check_cdbs,
    )


if not os.path.exists('debian/compat'):
    sys.exit(0)

# Package currently stores compat version in debian/compat..

debhelper_compat_version = read_debhelper_compat_file('debian/compat')

# debhelper >= 11 supports the magic debhelper-compat build-dependency.

# Exclude cdbs, since it only knows to get the debhelper compat version
# from debian/compat.

# debhelper-compat is only supported for stable compat levels
# https://bugs.debian.org/1026252

if (debhelper_compat_version < 11 or check_cdbs()
        or debhelper_compat_version > highest_stable_compat_level()):
    sys.exit(0)

# Upgrade to using debhelper-compat, drop debian/compat file.
os.unlink('debian/compat')

# Assume that the compat version is set in Build-Depends
with control as updater:
    insert_position = None
    changed_fields = []
    for field in ['Build-Depends', 'Build-Depends-Indep',
                  'Build-Depends-Arch']:
        to_delete: List[int] = []
        for offset, relation in iter_relations(
                updater.source.get(field, ''), 'debhelper'):
            if (field == 'Build-Depends' and
                    set([r.name for r in relation]) == set(['debhelper'])):
                # In the simple case, we'd just replace the debhelper
                # dependency with a debhelper-compat one, so remember the
                # location.
                insert_position = offset - len(to_delete)
            if is_relation_implied(
                    relation, 'debhelper (>= %d)' % debhelper_compat_version):
                to_delete.append(offset)

        if to_delete:
            # TODO(jelmer): Move this into a helper function in
            # lintian_brush.control.
            relations = parse_relations(updater.source[field])
            for i in reversed(to_delete):
                if i == 0 and len(relations) > 1:
                    # If the first item is removed, then copy the spacing to
                    # the next item
                    relations[1] = (
                        relations[0][0], relations[1][1], relations[0][2])
                del relations[i]

            updater.source[field] = format_relations(relations)
            changed_fields.append(field)

    updater.source["Build-Depends"] = ensure_exact_version(
        updater.source.get("Build-Depends", ""), "debhelper-compat",
        "%d" % debhelper_compat_version, position=insert_position)
    fixed_lintian_tag(updater.source, 'uses-debhelper-compat-file', ())

    for field in changed_fields:
        if updater.source.get(field) == "":
            del updater.source[field]

report_result("Set debhelper-compat version in Build-Depends.")

File: lintian-brush/src/fixers/uses_debhelper_compat_file.rs
use crate::{FixerError, FixerResult, LintianIssue};
use debian_analyzer::control::TemplatedControlEditor;
use debian_analyzer::debhelper::{highest_stable_compat_level, read_debhelper_compat_file};
use debian_analyzer::relations::is_relation_implied;
use debian_control::lossless::Entry;
use debversion::Version;
use std::path::Path;
use std::str::FromStr;

/// Check if the package uses CDBS by looking for cdbs include in debian/rules
fn check_cdbs(base_path: &Path) -> bool {
    let rules_path = base_path.join("debian/rules");
    if let Ok(content) = std::fs::read_to_string(&rules_path) {
        content.contains("/usr/share/cdbs/")
    } else {
        false
    }
}

pub fn run(base_path: &Path) -> Result<FixerResult, FixerError> {
    let compat_path = base_path.join("debian/compat");

    // Check if debian/compat exists
    let debhelper_compat_version = match read_debhelper_compat_file(&compat_path)? {
        Some(version) => version,
        None => return Err(FixerError::NoChanges),
    };

    // debhelper >= 11 supports the magic debhelper-compat build-dependency
    if debhelper_compat_version < 11 {
        return Err(FixerError::NoChanges);
    }

    // Exclude cdbs, since it only knows to get the debhelper compat version from debian/compat
    if check_cdbs(base_path) {
        return Err(FixerError::NoChanges);
    }

    // debhelper-compat is only supported for stable compat levels
    if debhelper_compat_version > highest_stable_compat_level() {
        return Err(FixerError::NoChanges);
    }

    let issue = LintianIssue::source_with_info(
        "uses-debhelper-compat-file",
        vec!["[debian/compat]".to_string()],
    );

    if !issue.should_fix(base_path) {
        return Err(FixerError::NoChangesAfterOverrides(vec![issue]));
    }

    // Parse and edit the control file
    let control_path = base_path.join("debian/control");
    let editor = TemplatedControlEditor::open(&control_path)?;

    let Some(mut source) = editor.source() else {
        return Err(FixerError::NoChanges);
    };

    // Create a target entry: debhelper (>= compat_version) for comparison
    let target_str = format!("debhelper (>= {})", debhelper_compat_version);
    let target_entry = Entry::from_str(&target_str)
        .map_err(|e| FixerError::Other(format!("Failed to parse target entry: {:?}", e)))?;

    // Process all three build dependency fields to remove debhelper
    for field in ["Build-Depends", "Build-Depends-Indep", "Build-Depends-Arch"] {
        let Some(mut deps) = (match field {
            "Build-Depends" => source.build_depends(),
            "Build-Depends-Indep" => source.build_depends_indep(),
            "Build-Depends-Arch" => source.build_depends_arch(),
            _ => None,
        }) else {
            continue;
        };

        // Check if this field has a debhelper dependency
        if !deps.has_relation("debhelper") {
            continue;
        }

        let Ok((_pos, entry)) = deps.get_relation("debhelper") else {
            continue;
        };

        // Only remove if the entry is implied by debhelper >= compat_version
        if !is_relation_implied(&entry, &target_entry) {
            continue;
        }

        // Remove debhelper
        deps.drop_dependency("debhelper");

        // Update or remove the field
        if deps.is_empty() {
            source.as_mut_deb822().remove(field);
        } else {
            match field {
                "Build-Depends" => source.set_build_depends(&deps),
                _ => source.set(field, &deps.to_string()),
            }
        }
    }

    // Add debhelper-compat to Build-Depends (Relations will sort appropriately)
    let mut build_depends = source.build_depends().unwrap_or_default();

    // Create the debhelper-compat entry
    let compat_version_str = debhelper_compat_version.to_string();
    let version = Version::from_str(&compat_version_str)
        .map_err(|e| FixerError::Other(format!("Failed to parse version: {:?}", e)))?;

    debian_analyzer::relations::ensure_exact_version(
        &mut build_depends,
        "debhelper-compat",
        &version,
        None,
    );

    source.set_build_depends(&build_depends);

    // Commit the control file changes
    editor.commit()?;

    // Delete the debian/compat file
    std::fs::remove_file(&compat_path)?;

    Ok(
        FixerResult::builder("Set debhelper-compat version in Build-Depends.")
            .fixed_issue(issue)
            .build(),
    )
}

crate::declare_fixer! {
    name: "uses-debhelper-compat-file",
    tags: ["uses-debhelper-compat-file"],
    apply: |basedir, _package, _version, _preferences| {
        run(basedir)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::builtin_fixers::BuiltinFixer;
    use std::fs;
    use tempfile::TempDir;

    #[test]
    fn test_simple() {
        let temp_dir = TempDir::new().unwrap();
        let debian_dir = temp_dir.path().join("debian");
        fs::create_dir_all(&debian_dir).unwrap();

        // Create debian/compat
        let compat_path = debian_dir.join("compat");
        fs::write(&compat_path, "11\n").unwrap();

        // Create debian/control
        let control_content = r#"Source: f2fs-tools
Section: admin
Priority: optional
Maintainer: Jelmer Vernooĳ <jelmer@debian.org>
Build-Depends:
 debhelper (>= 11),
 pkg-config,
 uuid-dev
Standards-Version: 4.2.0

Package: blah
Architecture: linux-any
Description: test
"#;
        let control_path = debian_dir.join("control");
        fs::write(&control_path, control_content).unwrap();

        let fixer = FixerImpl;
        let version: crate::Version = "1.0".parse().unwrap();
        let result = fixer.apply(temp_dir.path(), "f2fs-tools", &version, &Default::default());
        assert!(result.is_ok());

        // Check that compat file is deleted
        assert!(!compat_path.exists());

        // Check that control file is updated
        let updated_content = fs::read_to_string(&control_path).unwrap();
        assert!(updated_content.contains("debhelper-compat (= 11)"));
        assert!(!updated_content.contains("debhelper (>= 11)"));
    }

    #[test]
    fn test_no_change_when_compat_too_old() {
        let temp_dir = TempDir::new().unwrap();
        let debian_dir = temp_dir.path().join("debian");
        fs::create_dir_all(&debian_dir).unwrap();

        // Create debian/compat with old version
        let compat_path = debian_dir.join("compat");
        fs::write(&compat_path, "9\n").unwrap();

        // Create debian/control
        let control_content = r#"Source: test
Build-Depends: debhelper (>= 9)

Package: test
Description: test
"#;
        let control_path = debian_dir.join("control");
        fs::write(&control_path, control_content).unwrap();

        let fixer = FixerImpl;
        let version: crate::Version = "1.0".parse().unwrap();
        let result = fixer.apply(temp_dir.path(), "test", &version, &Default::default());
        assert!(matches!(result, Err(FixerError::NoChanges)));

        // Compat file should still exist
        assert!(compat_path.exists());
    }
}
